pysmt/cnf.py

```diff
@@ -0,0 +1,217 @@
+        DagWalker.__init__(self)
+        self.env = environment if environment else shortcuts.get_env()
+        self.mgr = self.env.formula_manager
+        self._introduced_variables = {}
+        self._cnf_pieces = {}
+
+    def _key_var(self, formula):
+        if formula in self._introduced_variables:
+            res = self._introduced_variables[formula]
+        else:
+            res = self.mgr.FreshSymbol()
+            self._introduced_variables[formula] = res
+        return res
+
+    def convert(self, formula):
+        tl, cnf = self.walk(formula)
+        res = [frozenset([tl])]
+        for clause in cnf:
+            if len(clause) == 0:
+                return CNFizer.FALSE_CNF
+            simp = []
+            for lit in clause:
+                if lit.is_true():
+                    simp = None
+                    break
+                elif not lit.is_false():
+                    simp.append(lit)
+            if simp:
+                res.append(frozenset(simp))
+        return frozenset(res)
+
+    def convert_as_formula(self, formula):
+        lsts = self.convert(formula)
+
+        conj = []
+        for clause in lsts:
+            if len(clause) == 0:
+                return self.mgr.FALSE()
+            elif len(clause) == 1:
+                conj.append(next(iter(clause)))
+            else:
+                conj.append(self.mgr.Or(clause))
+
+        if len(conj) == 0:
+            return self.mgr.TRUE()
+        elif len(conj) == 1:
+            return conj[0]
+        else:
+            return self.mgr.And(conj)
+
+    def printer(self, cnf):
+        print "{",
+        for clause in cnf:
+            print " {",
+            for lit in clause:
+                print "",lit,
+            print "}",
+        print " }"
+
+
+    def walk_forall(self, formula, args):
+        raise NotImplementedError("CNFizer does not support quantifiers")
+
+    def walk_exists(self, formula, args):
+        raise NotImplementedError("CNFizer does not support quantifiers")
+
+    def walk_and(self, formula, args):
+        if len(args) == 1:
+            return args[0]
+
+        k = self._key_var(formula)
+        cnf = [frozenset([k] + [self.mgr.Not(a).simplify() for a,_ in args])]
+        for a,c in args:
+            cnf.append(frozenset([a, self.mgr.Not(k)]))
+            for clause in c:
+                cnf.append(clause)
+        return k, frozenset(cnf)
+
+    def walk_or(self, formula, args):
+        if len(args) == 1:
+            return args[0]
+        k = self._key_var(formula)
+        cnf = [frozenset([self.mgr.Not(k)] + [a for a,_ in args])]
+        for a,c in args:
+            cnf.append(frozenset([k, self.mgr.Not(a)]))
+            for clause in c:
+                cnf.append(clause)
+        return k, frozenset(cnf)
+
+    def walk_not(self, formula, args):
+        a, cnf = args[0]
+        if a.is_true():
+            return self.mgr.FALSE(), CNFizer.TRUE_CNF
+        elif a.is_false():
```

Comment:
No, the contract of walk is the following: 
it returns a pair (v, cnf) such that the input formula is equivalent to v & cnf. v can be either a variable or a boolean constant and cnf is a set of clauses that are needed to explain v.

In this case I'm returning False, True that is logically false. The reason why I use TRUE_CNF instead of FALSE_CNF is that I need nothing to "explain" false. It would be incorrect to use FALSE_CNF. 

For this reason,